### PR TITLE
docs(lora): refute four #239 hypotheses; frozen-unit gathering becomes opt-in

### PR DESCRIPTION
**Pull request by saforem2**
## What this is

**Not a fix.** #239 is still open. This PR ships the investigation: four
refuted hypotheses, the real watchdog trace, and a lever — kept **off by
default** — that reproduces the decisive negative result.

The branch started out as a fix. The fix failed its own experiment on
Perlmutter, so it was inverted rather than quietly retained.

## The negative result

Keeping fully-frozen FSDP2 units gathered removes the 14/12 all-gather /
reduce-scatter asymmetry. That intervention provably landed — enqueued
ops drop 39 → 37, exactly the two removed all-gathers — and the deadlock
survived it unchanged.

| | baseline (`57601590`) | asymmetry removed (`57602201`) |
|---|---|---|
| stuck collective | `_REDUCE_SCATTER_BASE` | `_REDUCE_SCATTER_BASE` |
| payload | `419840 → 52480` | **identical** |
| SeqNum | 18 | 17 |
| PG status | enq 39 / started 19 / done 17 | enq 37 / started 18 / done 16 |
| r8 | `rc=134`, 3/3, `iter=NONE` | `rc=134`, `iter=NONE` |
| r16 | — | `rc=124`, `iter=NONE` |

Same skipped-work signature, renumbered by exactly the intervention's own
delta. So the asymmetry is **not the cause**, and — since the hang
persists without it — **not a precondition** either.

Baseline was run three times first, deliberately: two earlier jobs had
each hung once, which is equally consistent with a deterministic bug and
a coin flip. 3/3 settles it as deterministic, which every downstream
claim depends on.

## What changed

- `frozen_unit_kwargs()` is **inverted to opt-in**, `EZPZ_FSDP_FROZEN_RESHARD=1`
  (opt out) → `EZPZ_FSDP_KEEP_FROZEN_GATHERED=1` (opt in), default OFF.
  It costs ~**+1.7 GiB/rank** on `agpt-2b` at `world_size=8` bf16 and buys
  nothing; shipping a memory regression that failed its own experiment
  would be worse than shipping nothing. It stays in the tree so the
  negative result is re-runnable from a released build.
- `docs/guides/lora-fsdp-deadlock.md` — the full investigation. Leads
  with an open-bug danger box; documents **four** refuted hypotheses
  (LoRA-specificity, payload size, per-rank collective order, and now the
  asymmetry itself); ends with three live leads.
- `docs/troubleshooting.md` — **lowering** `TORCH_DDP_TIMEOUT` is the hang
  diagnostic. It defaults to 3600s, so in a 29-minute allocation the
  watchdog can never fire — several early probes were silent for that
  reason alone, and the silence was briefly misread as evidence. Every
  prior mention in the docs said to *raise* it.
- `tests/test_fsdp_tp_frozen_unit_reshard.py` — real `fully_shard` on 2
  gloo ranks via `mp.spawn`. Measures the asymmetry (14/12), measures the
  lever removing it (12/12), and guards that it stays **off by default**.
- `tests/test_tinker_lora_fsdp.py` — the real-model multi-rank probe
  behind refutation #3, previously uncommitted. Uses the actual
  `Transformer` + `apply_lora` and reproduces `parallelize`'s real
  grouping at ws=2 (at ws=1 FSDP2 issues *zero* collectives, so nothing
  about ordering is under test).
- `experiments/perlmutter/lora_239_repro.sbatch` — asserts the lever is
  live before running, so no arm can silently test the wrong build; each
  cell carries its recorded outcome inline.

## Known workaround for users

`--lora-rank 32` or higher completes normally. Why it does, with the
identical asymmetry and op sequence, is the sharpest unused clue and is
written up as lead #1.

## Verification

- **1762 passed**, 11 skipped. One pre-existing failure
  (`comprehensive_test.py::test_configs_module`, asserts
  `command_exists("python")`) deselected — verified via `git stash` that
  it fails on `main` too; this Mac has only `python3` on PATH.
- `ruff check` clean, no format churn.
- **Caught a CI-only bug in my own test.** torch 2.13 renamed the
  primitives FSDP2 calls (`all_gather_into_tensor` → `all_gather_single`),
  so the probe hooked nothing and counted 0 collectives — and `0 == 0` is
  *symmetric*, so the symmetry assertion would have passed vacuously.
  Fixed to patch whichever pair exists, assert a hook was installed, and
  pin absolute counts. Re-verified under **both** 2.12.1 and 2.13.0
  (14/12 → 12/12 on each). Local `.venv` is 2.12.1 and CI is 2.13, so
  this class of bug is invisible locally.
- **Mutation-tested.** Flipping the env default so the failed lever
  ships ON fails `test_frozen_unit_kwargs_is_off_by_default` and only
  that test. The spawn tests import the shipped helper rather than
  reimplementing it, so reverting `fsdp_tp.py` breaks them.
- Real hardware: Perlmutter 8× A100, torch 2.13.0+cu130, jobs
  `57601590` (baseline ×3) and `57602201` (intervention).

Closes nothing. #239 stays open.

## Summary by Sourcery

Document and reproduce the unresolved Perlmutter-only LoRA/FSDP2 deadlock while retaining the failed frozen-unit intervention as an opt-in experiment and hardening HPC validation and diagnostics.

New Features:
- Add an opt-in control for keeping fully frozen FSDP2 units gathered, preserving the released behavior by default.
- Add multi-rank FSDP2/LoRA probes, standalone reproduction material, and HPC experiment scripts for investigating issue #239.

Bug Fixes:
- Correct Polaris module loading to use currently available unpinned compiler and HDF5 module versions.
- Strengthen failover hostname validation to compare the recorded hostname field exactly.

Enhancements:
- Document the investigation into the Perlmutter-only LoRA/FSDP2 deadlock, including refuted causes, reproducibility evidence, remaining leads, and the known rank workaround.
- Add comprehensive HPC execution guidance covering environment setup, scheduler behavior, module pitfalls, experiment classification, and cluster-specific debugging.
- Improve hang diagnostics by documenting lower distributed timeouts and relevant NCCL tracing settings.
- Require least-privilege read-only repository permissions in the pytest workflow.

CI:
- Limit the pytest workflow GITHUB_TOKEN to read-only repository contents.

Documentation:
- Publish the LoRA/FSDP2 deadlock investigation and link it from the documentation navigation.
- Expand troubleshooting and Perlmutter guidance with hang diagnosis and environment failure handling.
- Add HPC operational guidance and draft support-ticket/reproducer notes.

Tests:
- Add real two-rank gloo tests covering frozen-unit collective asymmetry, opt-in behavior, default-off behavior, and real LoRA unit grouping.
- Add multi-rank tests verifying FSDP2 collective ordering and gradient routing across ranks, LoRA ranks, and target sets.
- Make collective instrumentation compatible with renamed torch 2.13 primitives and assert that hooks and absolute counts are valid.

Chores:
- Add repository guidance requiring HPC users to read the new operational skill guide first.

### CODE REVIEW

File: experiments/perlmutter/lora_239_collective_order.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+                      RANK=str(rank), WORLD_SIZE=str(ws))
+    torch.set_default_device("cpu")
+    dist.init_process_group("gloo", rank=rank, world_size=ws)
+    m = mk(r)
+    m.tok_embeddings.weight.requires_grad_(False)
+    m.output.weight.requires_grad_(False)
+    for p in m.norm.parameters():
+        p.requires_grad_(False)
+    seq = []; ph = ["fwd"]
+    oag, ors = dist.all_gather_into_tensor, dist.reduce_scatter_tensor
```
[the remark this answers is not in the document]
Comment (saforem2) [reply experiments/perlmutter/lora_239_collective_order.py]: Correct, and thanks — this is the same torch 2.13 rename bug I fixed in the two test files but missed here. It matters more than usual: this script produced refutation #3's evidence, and on 2.13 it recorded an **empty** sequence — empty strings compare equal across ranks, so the ordering check would have passed *vacuously* rather than failing.

Fixed in `d625f9b`: patches whichever of `all_gather_into_tensor`/`all_gather_single` (and the reduce-scatter pair) actually exists, and asserts both that a hook was installed and that the recorded sequence is non-empty.

File: experiments/perlmutter/lora_239_collective_order.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    def ag(*a, **k):
+        seq.append((ph[0], "A")); return oag(*a, **k)
+
+    def rs(*a, **k):
+        i = k.get("input", a[1] if len(a) > 1 else None)
+        seq.append((ph[0], "R", i.numel() if i is not None else -1)); return ors(*a, **k)
+
+    dist.all_gather_into_tensor = ag; dist.reduce_scatter_tensor = rs
+    mesh = init_device_mesh("cpu", (ws,)); kw = dict(mesh=mesh, reshard_after_forward=True)
```
[the remark this answers is not in the document]
Comment (saforem2) [reply experiments/perlmutter/lora_239_collective_order.py:83]: Agreed — same finding as the Codex thread above, and right. Fixed in `d625f9b`: patch whichever name pair exists, plus assertions that a hook landed and the trace is non-empty, so a future rename fails loudly instead of silently recording nothing.

File: experiments/perlmutter/lora_239_rank_bisect.sbatch [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+# and 57602201; r32/r64 reported OK), so spend the allocation entirely
+# on the unexplored interval. Order is binary-search, not ascending.
+for r in ${RANKS:-24 20 28}; do
+    probe "${r}"
+done
+
+echo
+echo "=== artifacts: ${OUT} ==="
+echo "=== VERDICT TABLE ==="
+grep -h "^### " "$0".log 2>/dev/null || true
```
[the remark this answers is not in the document]
Comment (saforem2) [reply experiments/perlmutter/lora_239_rank_bisect.sbatch]: Verified and fixed in `d625f9b`. `"$0".log` is the script path with `.log` appended, which never exists — the real log is `%x-%j.log` in the submit dir, and the job can't reliably name it from inside. Replaced with a pointer to the `### ` lines already echoed above.

File: experiments/perlmutter/lora_239_repro.sbatch [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+# RESULT (job 57602201): still hung. rc=134 (r8) / rc=124 (r16),
+# last_iter=NONE. Same collective, same 419840->52480 payload, merely
+# renumbered. This is a REFUTATION, kept runnable, not a fix.
+run "2-nofrozenreshard-r8"   1 --lora-rank 8
+run "3-nofrozenreshard-r16"  1 --lora-rank 16
+
+echo
+echo "=== artifacts: ${OUT} ==="
+echo "=== VERDICT TABLE ==="
+grep -h "^### " "${OUT}"/../lora-239-${SLURM_JOB_ID}.log 2>/dev/null || true
```
[the remark this answers is not in the document]
Comment (saforem2) [reply experiments/perlmutter/lora_239_repro.sbatch]: Same bug, confirmed empirically: this job printed an **empty** VERDICT TABLE on its actual run. Fixed in `d625f9b` the same way.

File: docs/guides/lora-fsdp-deadlock.md [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+    #239 is **unresolved**. The frozen-unit asymmetry described below is
+    real and measurable, but removing it on Perlmutter **did not stop the
+    deadlock** — see [Refuted: removing the asymmetry](#refuted-removing-the-asymmetry-fixes-it).
+    That intervention therefore ships **off by default**.
+
+    **Four** plausible-sounding explanations have now been tested and
+    refuted. They are documented here so nobody re-derives them.
+
+    Known workaround: use `--lora-rank 32` or higher, which completes
+    normally. Why it does is still unexplained.
```
[the remark this answers is not in the document]
Comment (saforem2) [reply docs/guides/lora-fsdp-deadlock.md]: Fair, and the scope got worse after I posted — the bisect has since moved the boundary from r>=32 to exactly 16→18. Every passing run is still one configuration (`agpt-2b`, `tp=1`, `bs=1`, `seq_len=2048`, `attn,mlp`, ws=8, torch 2.13 on A100), and with the mechanism unknown a different model or world size could move it.

Reworded in `d625f9b` to state it as measured-not-guaranteed, list the exact configuration, and ask users who hit the hang above r=18 to report it on #239.

File: tests/test_tinker_lora_fsdp.py [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+
+import pytest
+
+torch = pytest.importorskip("torch")
+
+pytestmark = pytest.mark.skipif(
+    sys.platform == "win32", reason="mp.spawn + gloo fixture is POSIX-only"
+)
+
+PORT = "29887"
```
[the remark this answers is not in the document]
Comment (saforem2) [reply tests/test_tinker_lora_fsdp.py]: Good catch — fixed in `d625f9b`. Binds `:0` in the parent, reads the OS-assigned port, and passes it to the workers. Verified by running this file and the sibling spawn file **concurrently**, which is the case you flagged: 7 + 5 pass.

File: tests/test_fsdp_tp_frozen_unit_reshard.py [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+import pytest
+
+torch = pytest.importorskip("torch")
+import torch.nn as nn  # noqa: E402
+
+pytestmark = pytest.mark.skipif(
+    sys.platform == "win32", reason="mp.spawn + gloo fixture is POSIX-only"
+)
+
+PORT = "29874"
```
[the remark this answers is not in the document]
Comment (saforem2) [reply tests/test_fsdp_tp_frozen_unit_reshard.py]: Same fix in `d625f9b` (free port chosen in the parent, passed into `_worker`), verified with both spawn files running concurrently.

**Comment by saforem2** [comment]
@sourcery-ai review

**Comment by saforem2** [comment]
## Merge note: this cuts **v0.27.4**

Heads-up before merging — with `release:patch`, `release-on-merge.yml`
fires on PR close and publishes a real tagged release (0.27.3 → 0.27.4),
not just a branch landing. Flagging it because most of this PR is docs
and experiment scripts, so it may not read like a release.

**What actually ships to users** (3 source files, +82/-8):

- `src/ezpz/bin/utils.sh` — **the only behavior fix.** Unpins
  `gcc-native` and `cray-hdf5-parallel` in `ezpz_load_modules_polaris`;
  Polaris removed the pinned versions, so module loading was broken
  there for everyone.
- `src/ezpz/examples/fsdp_tp.py` — adds `frozen_unit_kwargs`, **opt-in
  and off by default**. With `EZPZ_FSDP_KEEP_FROZEN_GATHERED` unset the
  runtime path is byte-identical to `main`, so there is no behavior
  change for existing users. It is retained only so the failed #239
  experiment stays reproducible from a released build.
- `src/ezpz/tinker/lora.py` — docstring correction.

Everything else is docs (the #239 investigation, `SKILL.md`), experiment
scripts, tests, and the two CodeQL fixes.

So `release:patch` is the right label: a genuine bug fix for Polaris
users, no API or default-behavior change. If you would rather not cut a
version for this, relabel `release:skip` before merging.

**Comment by saforem2** [comment]
@sourcery-ai review

**Comment by saforem2** [comment]
Superseded by a split, so automated review can actually run:

- **#241** — code only (`src/` + `tests/` + the CodeQL workflow fix), **43.6k** diff chars, `release:patch`
- **#242** — docs + experiment scripts, stacked on #241, `release:skip`

**Why:** this PR is **215k** diff chars and Sourcery hard-refuses anything over 150k ("your pull request is larger than the review limit of 150,000 diff characters"). So the fresh bot round I was waiting on was never going to arrive, and 68 commits were about to merge with no bot review at all — including `utils.sh`, which has already carried two bugs I introduced and then found by accident.

Same content, same commits' worth of change, verified identical file-by-file against this branch. Merge #241 first, then #242.

Closing in favour of those two.
